pkg/pgmodel/migrate.go

```diff
@@ -26,133 +28,344 @@ const (
+	migrateMutex.Lock()
+	defer migrateMutex.Unlock()
+
+	// Getting an early connection to install the extra extension.
+	// TODO: Investigate why this is required. Installing the extension on the
+	// migration connection fails, thats why this is necessary. This should be removed.
+	conn, err := db.Acquire(context.Background())
 	if err != nil {
-		return r, err
+		return fmt.Errorf("could not acquire connection: %w", err)
 	}
-	err = r.Close()
+
+	defer conn.Release()
+
+	appVersion, err := semver.Make(versionInfo.Version)
 	if err != nil {
-		return r, err
+		return fmt.Errorf("app version is not semver format, aborting migration")
+	}
+	if err := ensureVersionTable(db); err != nil {
+		return fmt.Errorf("error ensuring version table: %w", err)
+	}
+	dbVersion, err := getDBVersion(db)
+	if err != nil {
+		return fmt.Errorf("failed to get the version from database: %w", err)
+	}
+
+	// If already at correct version, nothing to migrate.
+	if dbVersion.Compare(appVersion) == 0 {
+		return nil
 	}
-	s := buf.String()
-	s = strings.ReplaceAll(s, "SCHEMA_CATALOG", catalogSchema)
-	s = strings.ReplaceAll(s, "SCHEMA_EXT", extSchema)
-	s = strings.ReplaceAll(s, "SCHEMA_PROM", promSchema)
-	s = strings.ReplaceAll(s, "SCHEMA_SERIES", seriesViewSchema)
-	s = strings.ReplaceAll(s, "SCHEMA_METRIC", metricViewSchema)
-	s = strings.ReplaceAll(s, "SCHEMA_DATA", dataSchema)
-	s = strings.ReplaceAll(s, "SCHEMA_DATA_SERIES", dataSeriesSchema)
-	s = strings.ReplaceAll(s, "SCHEMA_INFO", infoSchema)
-	r = ioutil.NopCloser(strings.NewReader(s))
-	return r, err
-}
 
-func (t *mySrc) ReadUp(version uint) (r io.ReadCloser, identifier string, err error) {
-	r, identifier, err = t.Driver.ReadUp(version)
+	// Error if at a greater version.
+	if dbVersion.Compare(appVersion) > 0 {
+		return fmt.Errorf("schema version is above the application version, cannot migrate")
+	}
+
+	tx, err := db.Begin(context.Background())
 	if err != nil {
-		return
+		return fmt.Errorf("unable to start transaction: %w", err)
 	}
-	r, err = t.replaceSchemaNames(r)
-	return
-}
+	defer func() {
+		_ = tx.Rollback(context.Background())
+	}()
 
-func (t *mySrc) ReadDown(version uint) (r io.ReadCloser, identifier string, err error) {
-	r, identifier, err = t.Driver.ReadDown(version)
+	_, err = tx.Exec(context.Background(), timescaleInstall)
 	if err != nil {
-		return
+		return fmt.Errorf("timescaledb failed to install due to %w", err)
 	}
-	r, err = t.replaceSchemaNames(r)
-	return
-}
 
-func metadataUpdate(db *sql.DB, withExtension bool, key string, value string) {
-	/* Ignore error if it doesn't work */
-	if withExtension {
-		_, _ = db.Exec(metadataUpdateWithExtension, key, value, true)
+	// No version in DB.
+	if dbVersion.Compare(semver.Version{}) == 0 {
+		if err = execMigrationFiles(tx, preinstallScripts); err != nil {
+			return err
+		}
 	} else {
-		_, _ = db.Exec(metadataUpdateNoExtension, key, value, true)
+		if err = upgradeVersion(tx, dbVersion, appVersion); err != nil {
```

Comment:
@ante let's put an error before this line for now (temporarily), because this isn't tested yet. Otherwise after you put in the WARNINGS @JLockerman suggested plus swap the order of the two replace lines, it's good to merge. Does not need another review before merge